QgsEditorWidgetRegistry disconnect signals when mapLayer will be removed

rldhont opened the pull request:
Add a method to disconnect signals added when layerIsAdded to QgsMapLayerRegistry. This method is connect to layerWillBeRemoved.

This method will probably fix an issue in QGIS Server about performance deterioration.
Each time a layer is added to QgsMapLayerRegistry by QGIS Server, QgsEditorWidgetRegistry connects to appropriate signals from map layers to load and save style but NEVER disconnects its.

rldhont committed:
QgsEditorWidgetRegistry disconnect signals when mapLayer will be removed

Add a method to disconnect signals added when layerIsAdded to QgsMapLayerRegistry. This method is connect to layerWillBeRemoved.

This method will probably fix an issue in QGIS Server about performance deterioration.
Each time a layer is added to QgsMapLayerRegistry by QGIS Server, QgsEditorWidgetRegistry connects to appropriate signals from map layers to load and save style but NEVER disconnects its.

rldhont commented:
@dmarteau can you test it in QGIS Server Context ?

m-kuhn commented:
I assumed that map layers will always be deleted when they are removed.
Is this not the case and the layers are recycled?

dmarteau commented:
Yes, I' m backporting the patch into our test branch (based on 2.14)

> Le 8 juin 2016 à 20:06, rldhont <notifications@github.com> a écrit :
> 
> @dmarteau <https://github.com/dmarteau> can you test it in QGIS Server Context ?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/qgis/QGIS/pull/3186#issuecomment-224677786>, or mute the thread <https://github.com/notifications/unsubscribe/ABvOTtiN6PDEBXJcRIw0wwJatsWJpgKGks5qJwSTgaJpZM4IxPTA>.
>

rldhont commented:
Hi @m-kuhn,

In the QGIS Server context, to avoid time consuming by layer creation from xml, layers are cached. So layers are never been removed and are recycled at each request.

For example, at the first request, the layer is created, added to the cache, added to QgsMapLayerRegistry to be used, and then removed from QgsMapLayerRegistry. At the second and the others request, layers are picked from the cache, added to QgsMapLayerRegistry to be used, and then removed fromQgsMapLayerRegistry. So connections are added at each request and never been removed. It is why I proposed this fix.

NathanW2 commented:
Going into 3.0 I would like to see us let the renderer code take layer
objects so we don't need to keep using the global for a one off renders,
just make the renderer, pass in the layers, done

On Thu, 9 Jun 2016 5:28 pm rldhont <notifications@github.com> wrote:

> Hi @m-kuhn <https://github.com/m-kuhn>,
>
> In the QGIS Server context, to avoid time consuming by layer creation from
> xml, layers are cached. So layers are never been removed and are recycled
> at each request.
>
> For example, at the first request, the layer is created, added to the
> cache, added to QgsMapLayerRegistry to be used, and then removed from
> QgsMapLayerRegistry. At the second and the others request, layers are
> picked from the cache, added to QgsMapLayerRegistry to be used, and then
> removed fromQgsMapLayerRegistry. So connections are added at each request
> and never been removed. It is why I proposed this fix.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/3186#issuecomment-224819736>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAXS3KNJNMENjmVDkD9qfurnbSfbmuiXks5qJ8CsgaJpZM4IxPTA>
> .
>

m-kuhn commented:
> Going into 3.0 I would like to see us let the renderer code take layer

Yes, the layer registry should really be used as little as possible. But for now, this patch looks good to me. Thanks.

dmarteau commented:
I Have tested the patch:  it does not work in the context of qgis server  because QgsEditorWidgetRegistry record every layer added to the
QgsMapLayerRegistry by listening to signal 'layerWasAdded' but layerWillBeRemoved is emitted only for layers owned by
QgsMapLayerRegistry and thus layerWillBeRemoved is never called with qgis server.

In order for this patch to work you need to replace layerWillBeRemoved by layersWillBeRemoved which is emitted for
all layers added to QgsMapLayerRegistry.

I have made the fix and now  layerWillBeRemoved  is correctly called in QgsEditorWidgetRegistry and it solve the problem of cumulative performance degradation observed on qgis-server requests
by deconnecting from the added layers, so the diagnostic was good.



> Le 9 juin 2016 à 09:33, Matthias Kuhn <notifications@github.com> a écrit :
> 
> Going into 3.0 I would like to see us let the renderer code take layer
> 
> Yes, the layer registry should really be used as little as possible. But for now, this patch looks good to me. Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/qgis/QGIS/pull/3186#issuecomment-224820736>, or mute the thread <https://github.com/notifications/unsubscribe/ABvOTieWZVq32mcSP-OuQELCBvO2vitnks5qJ8HTgaJpZM4IxPTA>.
>

dmarteau commented:
@m-kuhn 
I have made a PR #3189, could review it please ?
